### PR TITLE
[project-base] Hack phpunit to make tests clean

### LINKED ISSUE #907: Services needed in integration tests should not be manually registered as public
### What is happening

It would be best if we wouldn't have to make every service that is taken from DIC in tests public. 

These copy-pasted service definitions can cause a lot of trouble. Either stop using the container directly (can we autowire services in integration tests?) or have all services public in test env by default (maybe via a compiler pass?)...

### Expected result
We don't need to manually add services into services_test.yml.

### PR DESCRIPTION
**Pull request by simara-svatopluk**
| Q             | A
| ------------- | ---
|Description, reason for the PR| Services in tests are loaded by service locator, and that is ugly + anti dependency injection + must be annotated for phpstan + must utilize `services_test.yml`
|New feature| :x: 
|[BC breaks](https://github.com/shopsys/shopsys/blob/master/docs/contributing/backward-compatibility-promise.md)| No 
|Fixes issues| ...
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| :heavy_check_mark: 

Current problems
- [x] PHPStan suddenly understands the code more, so it discovered `dynamic field declaration` in `CartItemTest`, `CartTest`, `CartWatcherTest`. Remoted dynamic assigning fields that should't make any sense
- [x] The `Zalas\Injector` tool seems to require interfaces, not extended classes from project base
- [x] `ElasticsearchStructureUpdateCheckerTest` - needs to get `Domain` from container, cannot be injected, no idea why
- [x] `ProductOnCurrentDomainFacadeCountDataTest` - needs to get `ParameterRepository` from container, cannot be injected, no idea why
- [x] - services from framework must be mentioned in `services_tests.yml` anyway :-(

### CODE REVIEW

**Comment by simara-svatopluk** [comment]
Hi @vitek-rostislav,

I found troubles in existing tests during implementation. Eg. there are product data, and `price` is setted `$productData->price = 100;` https://github.com/shopsys/shopsys/blob/master/project-base/tests/ShopBundle/Functional/Model/Cart/CartTest.php#L38 But the `ProductData` class doesn't contain such property.

What does it do? What shall it do? This trouble blocks building the project :-(

**Comment by simara-svatopluk** [comment]
@vitek-rostislav So after a day of coding, the work is done. You can find unsolved problems in description.

Tests are much cleaner, they look like regular classes, just nice result; PHPStan understands the code better; there is almost the same amount of code; and you don't have to register project-base classes in tests.

I didn't care about the history, I may fix it after you confirm code is changed correctly.
There is a lot of changes, I recommend to take a look on it fast before there is a ton of conflicts ;-)

**Comment by simara-svatopluk** [comment]
Important note - we cannot use `**DataInterface` in functional tests anymore as there is a new annotation fixer that "fixes" interface -> shopsys class. So the solution is a bit different than it was in https://github.com/shopsys/shopsys/issues/907#issuecomment-487841814

**Comment by PetrHeinz** [comment]
@simara-svatopluk That looks nice! :+1: Thanks!

As for the problem with extended classes you mention - have you tried running `php phing annotations fix` from #1344 that @vitek-rostislav prepared? It should replace all annotations of classes that are extended in the project repository.

In some cases an inline annotation such as `/** @var /Shopsys/ShopBundle/Model/Product/ProductData $productData */` is necessary. See [Framework Extensibility](https://github.com/shopsys/shopsys/blob/bd0ccc5acafbfa7b16ba3a24303480fd3c09269f/docs/introduction/framework-extensibility.md#problem-3) for description of such issue.

**Comment by simara-svatopluk** [comment]
Hi @PetrHeinz,

that's the point of my previous comment - this phing target (I run it transitively by `standards-fix`) "fixes" interface to project-base class. And therefore I'm unable to use `**Interface` mostly, and therefore the solution is a bit different than proposed in https://github.com/shopsys/shopsys/issues/907#issuecomment-487841814 :-)

Yep, I know sometimes it is necessary, but at least I found some places where it can be omitted, because there is now a concrete factory class instead of interface.

**Comment by LukasHeinz** [comment]
Hi @simara-svatopluk , thank you for your pull request!

I have noticed that you didn't sign contributor license agreements (I cannot find your GitHub name there). Before we can process your contribution you have to agree and sign this because there is no longer any contract between you and Shopsys.

The agreement is pretty straight forward - you just confirm that it is your idea and you give us the right to use your contrition (please read the agreements for full info). We need this even for such a small contribution like this. But you just have to sign it once, so next time you do not have to worry about that.

Thank you.

**Comment by simara-svatopluk** [comment]
@LukasHeinz You are right! I used to use different email when I worked for You, so ok, signed :heavy_check_mark:

**Comment by vitek-rostislav** [comment]
Hi @simara-svatopluk, thanks for the PR! :slightly_smiling_face: I am sorry for not responding to your [question](https://github.com/shopsys/shopsys/issues/907#issuecomment-528723137), I haven't checked github for a few days...

I see the PR is targeted to master, however, if there were BC breaks, it should be targeted to `9.0`...but do not worry, we can manage this ourselves - also, if the breaking changes are just the removals of services from `services_test.yml` (which causes the tests failing on current projects = BC break), you could just mark them as `@deprecated` - so no BC break then. The changes in `project-base` are excluded from BC promise - so the changes in projects's tests are not BC breaks.

> I found troubles in existing tests during implementation. Eg. there are product data, and price is setted $productData->price = 100;
...What shall it do?

I see you already fixed this in a separate commit, fine with me :+1: 

>  this phing target (I run it transitively by standards-fix) "fixes" interface to project-base class. And therefore I'm unable to use **Interface mostly

I still do not get it - you can still typehint the interfaces and the new fixer just changes the annotations of the given property to the project implementation. Could you please show me a particular place where this behavior is a problem?

Also, we have just started a new sprint, but sadly, this PR is not planned to be processed due to the other priorities. But don't worry, we will get to it sooner or later :wink:

**Comment by simara-svatopluk** [comment]
Hi @vitek-rostislav 
If you can manage the change by yourself, that's great, I don't care whether it will be targeted to `9.0` with BC break or to `master` with deprecation.

> you can still typehint the interfaces and the new fixer just changes the annotations of the given property to the project implementation

That is the point. I "can" use interfaces, but there are project-base implementations, so the fixer forces to use project-base implementation. So I **can not** use FW interfaces. And FW interfaces were used in the original solution. So the solution is a bit different. That's the only point - the solution is a bit different than the "prototyped solution".

I'm afraid that after two weeks, there will be a lots of changes in the code, so the rebase will be complicated. But you promised that You'll process this PR, so I'm counting on you :-)

**Comment by simara-svatopluk** [comment]
bump @vitek-rostislav what about now? Are you willing to review this improvement? ;-)

**Comment by vitek-rostislav** [comment]
Hi @simara-svatopluk, well as always, it is about priorities and planning - I am pushing this as the top priority, hopefully, it will be planned to the sprint that starts the day after tomorrow :wink:

**Comment by simara-svatopluk** [comment]
Hi @vitek-rostislav, how do you do? :-)
Did you include this PR to the current sprint?

**Comment by vitek-rostislav** [comment]
Hi @simara-svatopluk, sadly, it is not included in the sprint due to the other priorities :disappointed: Hopefully next time...I know I promised it will be processed but at the moment, there is nothing more I can do about it

**Comment by simara-svatopluk** [comment]
@vitek-rostislav This PR is sort of simple, but long. It changes mostly just tests....
So cannot you just spend like 10 minutes by looking at non-test files and then quickly go through tests... And if tests passes, everything should be fine. So at most this will take you 30 minutes, and if you'll be quick, then you can do it in 15 minutes.

And once you see it's ok, then just merge it. It's not difficult change and won't cause problems in production code at all.

**Comment by vitek-rostislav** [comment]
I am sorry @simara-svatopluk, I have literally no time for that :disappointed: And even if I had, you know it does not work this way - I can not merge it until someone else sees it, and until validation is done with @LukasHeinz.

**Comment by henzigo** [comment]
> Hi Svaťa, well I want it 😁 We will process it, you have my word 😉

Not working well

**Comment by s3tezsky** [comment]
Hello @simara-svatopluk, thanks for this PR! I walked through it and have some point which will be nice to get improved.

1. > I found troubles in existing tests during implementation. Eg. there are product data, and price is setted $productData->price = 100; https://github.com/shopsys/shopsys/blob/master/project-base/tests/ShopBundle/Functional/Model/Cart/CartTest.php#L38 But the ProductData class doesn't contain such property.

    This should be solved in aa different PR as it has nothing to do with "Hacking" a PHPUnit. I've created a new PR solving this (#1459). Could you please remove it from from this one?

2. Latest `zalas/phpunit-injector` is in a version `1.3.2`. Is there any reason why you are using `v1.2`?

3. Could you please rebase the branch to actual master? I've done it localy and there are no conflicts, but I would like to keep it actual in your fork.

---
I will do even more digging tomorrow. Anyway I like this PR and belive we will be able to merge it soon. 🚀

**Comment by simara-svatopluk** [comment]
Hi, thanks for review! I'm horribly busy this week, so I'll do it next week.
Quick answers:
1. sure
2. No particular reason, will use current
3. sure
4. yes

File: project-base/tests/ShopBundle/Functional/Model/Product/ProductOnCurrentDomainSqlFacadeCountDataTest.php [hunk 1]
```diff
@@ -4,16 +4,21 @@
...
-use Shopsys\FrameworkBundle\Model\Product\ProductOnCurrentDomainFacade;
 use Shopsys\FrameworkBundle\Model\Product\ProductOnCurrentDomainFacadeInterface;
 
 class ProductOnCurrentDomainSqlFacadeCountDataTest extends ProductOnCurrentDomainFacadeCountDataTest
 {
+    /**
+     * @var \Shopsys\FrameworkBundle\Model\Product\ProductOnCurrentDomainFacade
+     * @inject
+     */
+    protected $productOnCurrentDomainFacade;
```
Comment (pesektomas) [inline project-base/tests/ShopBundle/Functional/Model/Product/ProductOnCurrentDomainSqlFacadeCountDataTest.php]: Couldn't it be private?

File: project-base/tests/ShopBundle/Functional/Model/Payment/PaymentTest.php [hunk 2]
```diff
@@ -4,35 +4,46 @@
...
+
     public function testRemoveTransportFromPaymentAfterDelete()
     {
-        /** @var \Shopsys\ShopBundle\Model\Payment\PaymentDataFactory $paymentDataFactory */
-        $paymentDataFactory = $this->getContainer()->get(PaymentDataFactoryInterface::class);
-        /** @var \Shopsys\ShopBundle\Model\Transport\TransportDataFactory $transportDataFactory */
-        $transportDataFactory = $this->getContainer()->get(TransportDataFactoryInterface::class);
         $em = $this->getEntityManager();
 
         $vatData = new VatData();
```
[the remark this answers is not in the document]
Comment (s3tezsky) [reply project-base/tests/ShopBundle/Functional/Model/Payment/PaymentTest.php]: That is a good point. I do not want to do it in this PR. Could you create an issue or fix it in separate PR.

[the remark this answers is not in the document]
Comment (pesektomas) [reply project-base/tests/ShopBundle/Functional/Model/Payment/PaymentTest.php]: Yes, I will do it

File: project-base/tests/ShopBundle/Functional/Form/Admin/AdvancedSearch/AdvancedSearchOperatorTranslationTest.php [hunk 3]
```diff
@@ -5,31 +5,40 @@
...
 class AdvancedSearchOperatorTranslationTest extends FunctionalTestCase
 {
+    /**
+     * @var \Shopsys\FrameworkBundle\Model\AdvancedSearch\ProductAdvancedSearchConfig
+     * @inject
+     */
+    private $productAdvancedSearchConfig;
+
+    /**
+     * @var \Shopsys\FrameworkBundle\Model\AdvancedSearch\OrderAdvancedSearchConfig
```
[the remark this answers is not in the document]
Comment (s3tezsky) [reply project-base/tests/ShopBundle/Functional/Form/Admin/AdvancedSearch/AdvancedSearchOperatorTranslationTest.php]: This is definitely weird. I guess it was implemented wrong even before. It only proofes a weird quality of these test. With both services the tests are passing. 😕

**Comment by simara-svatopluk** [comment]
@simara-svatopluk approves too (cannot officially approve my own PR)

APPROVE

Guys, thank You very much for the effort, that's amazing!

**Comment by TomasVotruba** [comment]
Just curious... how did you automate this migration? 
Please don't tell me it was manual work

**Comment by s3tezsky** [comment]
Hi @TomasVotruba, I will not tell you when you rather don't want to know. Sadly, we do not have any automated migration.

**Comment by TomasVotruba** [comment]
Too bad. I hope you'll make more research next time, this is 1-2 hours job.

**Comment by TomasVotruba** [comment]
Finally, it's in @rectorphp

I overestimated my skill, it took me 3 hours in the end.

## Use

```bash
composer require rector/rector -dev
composer require zalas/phpunit-injector --dev
vendor/bin/rector process tests --set phpunit-injector
```

**Comment by simara-svatopluk** [comment]
That's amazing. I spent 8 hours on manual refactoring, and I guess colleagues spent even more during reviews/fixing.

You are the author of the Rector of course. But I think even if it would have taken me 5x more time (learn and try), it would have been 15 hours in total, I would have learn the Rector, and colleagues would have had spend less time.... And in the end I would have known a new tool.

I'll give the Rector a try next time.


And I have a suggestion for @s3tezsky - you've written upgrade instructions for this PR, you can change them into a single line `vendor/bin/rector process tests --set phpunit-injector` ;-)

**Comment by TomasVotruba** [comment]
> That's amazing. I spent 8 hours on manual refactoring, and I guess colleagues spent even more during reviews/fixing.

:cry: 


> But I think even if it would have taken me 5x more time (learn and try), it would have been 15 hours in total, I would have learn the Rector, and colleagues would have had spend less time.... And in the end I would have known a new tool.

I don't know any tool which is faster to learn than use at first :D took me 6 months to learn AST propperly.

> I'll give the Rector a try next time.

If you get stuck, call me B-)

**Comment by s3tezsky** [comment]
Hello guys,

@TomasVotruba really thanks for your effort and showing us a way how to work with rector! Sadly I cannot run it properly (see bellow).

@simara-svatopluk unfortunately we cannot put it in upgrade notes as an alternative because I do not know the way how to run it. All my tries ended up with a failure.

We cannot require it as a dev dependency because of conflicts of many outdated dependencies we are using.

Also I tried run it via docker image  with no succes.
```
$ docker run -v $(pwd):/project rector/rector process --set phpunit-injector --autoload-file /project/vendor/autoload.php /project/tests
Rector No version set (parsed as 1.0.0)@

   0/462 [░░░░░░░░░░░░░░░░░░░░░░░░░░░░]   0%
Fatal error: Class PHPStan\Type\StaticType contains 3 abstract methods and must therefore be declared abstract or implement the remaining methods (PHPStan\Type\Type::isArray, PHPStan\Type\Type::inferTemplateTypes, PHPStan\Type\Type::traverse) in /project/vendor/phpstan/phpstan/src/Type/StaticType.php on line 12
```

**Comment by TomasVotruba** [comment]
Yeah, we need to solve this to avoid conflicts
https://github.com/rectorphp/rector/issues/177

**Pull request merged**